mautic/mautic
# Enhances filters evaluation for DWCs

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [✖️]
| New feature/enhancement? (use the a.x branch)      | [✔️ ]
| Deprecations?                          | [✖️ ]
| BC breaks? (use the c.x branch)        | [✖️ ]
| Automated tests included?              | [✖️ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

Leuchtfeuer.com

This enhancement will allow to add another filters for your dynamic web content by listening to DynamicContentEvents::ON_CONTACTS_FILTER_EVALUATE event. To add new filter, subscriber of wished filter has to listen to Mautic\LeadBundle\LeadEvents::LIST_FILTERS_CHOICES_ON_GENERATE event and store a choice in LeadListFiltersChoicesEvent, in LeadBundle\EventListener\FilterOperatorSubscriber is an example how could it be implemented.

(this enhancement works for this moment only for landing pages. I want firstly get some feedback, if PR is good I will commit changes for email bundle too)

As requested, here is an example how could be used such enhancement:

In this plugin is possible to filter DWC by device type of page's visitor:
https://github.com/Leuchtfeuer/mautic-DwcDeviceType-bundle 
Steps to test this:
1. In /plugins/ directory clone this plugin
2. Rename plugin name from mautic-DwcDeviceType-bundle to LeuchtfeuerDwcDeviceTypeBundle(otherwise it's not going to work)
3. Clear the cache 
4. Enble plugin if it's not
5. Create new DWC which one is not campaign based and choose as filter - device type
6. Set this DWC in landing page by slot name

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
7. Add some plugin which listens to ON_CONTACTS_FILTER_EVALUATE event and enable this
8. Create DWC which is NOT campaign based(set a slot name) and add filters that you need
9. Place DWC into your landing page by setting in your page builder {dwc=<slot_name>}
